File: two_pointers/3_sum/main.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> triplets = {};

        sort(nums.begin(), nums.end());

        for(int i = 0; i < nums.size(); i++) {
            if(i > 0 && nums[i-1] == nums[i]) {
                continue;
            }

            int l = i + 1;
            int r = nums.size() - 1;

            while(l < r) {
                int res = nums[i] + nums[l] + nums[r];

                if (res == 0) {
                    triplets.push_back(vector<int>{nums[i], nums[l], nums[r]});
                    l++;

                    while (nums[l-1] == nums[l]) {
                        l++;
                    }
                } else if (res > 0) {
                    r--;
                } else {
                    l++;
                }
            }
        }

        return (triplets);
    }
};

File: two_pointers/3_sum/main.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()

        triplets = set()
        for i in range(len(nums)):
            l = i + 1
            r = len(nums) - 1

            while l < r:
                res = nums[i] + nums[l] + nums[r]
                if res == 0:
                    triplets.add((nums[i], nums[l], nums[r]))
                    l += 1
                    r -= 1
                elif res > 0:
                    r -= 1
                else:
                    l += 1
        return list(triplets)
